python/tank/platform/environment.py

```diff
@@ -878,3 +880,262 @@ def copy_apps(self, src_engine_name, dst_engine_name):
+            )
+
+            # update the settings by adding or removing keys based on the
+            # type of dumping being performed.
+            update_settings(engine_schema, engine_settings, engine_name,
+                engine_manifest_file, debug)
+
+            # processing all the installed apps
+            for app_name in self.get_apps(engine_name):
+
+                # only process settings in this file
+                (tokens, app_file) = self.find_location_for_app(engine_name,
+                    app_name)
+                if not app_file == self._env_path:
+                    continue
+
+                # drill down into the yml data to find the chunk where the
+                # app settings live
+                app_settings = yml_data
+                for token in tokens:
+                    app_settings = app_settings.get(token)
+
+                # get information about the app in order to process the
+                # settings.
+                app_descriptor = self.get_app_descriptor(engine_name, app_name)
+                app_schema = app_descriptor.get_configuration_schema()
+                app_manifest_file = os.path.join(app_descriptor.get_path(),
+                    constants.BUNDLE_METADATA_FILE)
+
+                # update the settings by adding or removing keys based on the
+                # type of dumping being performed.
+                update_settings(app_schema, app_settings, engine_name,
+                    app_manifest_file, debug)
+
+        # processing all the frameworks
+        for fw_name in self.get_frameworks():
+
+            # only process settings in this file
+            (tokens, fw_file) = self.find_location_for_framework(fw_name)
+            if not fw_file == self._env_path:
+                continue
+
+            # drill down into the yml data to find the chunk where the
+            # framework settings live
+            fw_settings = yml_data
+            for token in tokens:
+                fw_settings = fw_settings.get(token)
+
+            # get information about the framework in order to process the
+            # settings.
+            fw_descriptor = self.get_framework_descriptor(fw_name)
+            fw_schema = fw_descriptor.get_configuration_schema()
+            fw_manifest_file = os.path.join(fw_descriptor.get_path(),
+                constants.BUNDLE_METADATA_FILE)
+
+            # update the settings by adding or removing keys based on the
+            # type of dumping being performed.
+            update_settings(fw_schema, fw_settings, fw_manifest_file, debug)
+
+        # write the modified data to the output path
+        self.__write_data(self._env_path, yml_data)
+
+
+class UpdateAllSettingsFormat(object):
+    """
+    Format enumeration to use when updating all settings in an environment.
+
+    FULL = Include all settings, using default values when necessary.
+    SPARSE = Exclude settings using default values.
+    """
+    (FULL, SPARSE) = range(2)
```

Comment:
It was really only needed by the WritableEnvironment. I've moved it into that class and removed it from the validate action code.